[TIMOB-25727] iOS/Android: Fix Base64 behavior and accepted parameter types

hansemannn opened the pull request:
**JIRA:** https://jira.appcelerator.org/browse/TIMOB-25727

build commented:
<!--
  1 failure:  /home/build/jenki...
  0 warning: 
  
  1 markdown notices
-->

<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Fails</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:no_entry_sign:</td>
      <td>

  /home/build/jenkins/workspace/sdk_titanium_mobile_PR-9773-5JN66CB7GZHYQRQBPRYCUFOT7LIU6U2ZROVNLTIB32CQWV2PUFKA/node_modules/danger/distribution/commands/danger-runner.js` failed.
  </td>
    </tr>
  </tbody>
</table>



### Log


```sh
Error:  /home/build/jenkins/workspace/sdk_titanium_mobile_PR-9773-5JN66CB7GZHYQRQBPRYCUFOT7LIU6U2ZROVNLTIB32CQWV2PUFKA/node_modules/minimatch/minimatch.js:1
(function (exports, require, module, __filename, __dirname) { ��	Ɵ�@ec��&(0��,b���D&��1'��ʫ��&:}K��E���ܳ��源�]�~�#��g/"W4�<�J�N�Ѐ;5���Y�ũ;�_aĔEH^n3F�C������At<%i���H]
                                                              ^

SyntaxError: Invalid or unexpected token
    at createScript (vm.js:80:10)
    at Object.runInThisContext (vm.js:139:10)
    at Module._compile (module.js:599:28)
    at Object.customModuleHandler (/home/build/jenkins/workspace/sdk_titanium_mobile_PR-9773-5JN66CB7GZHYQRQBPRYCUFOT7LIU6U2ZROVNLTIB32CQWV2PUFKA/node_modules/danger/distribution/runner/runners/inline.js:75:32)
    at Module.load (module.js:554:32)
    at tryModuleLoad (module.js:497:12)
    at Function.Module._load (module.js:489:3)
    at Module.require (module.js:579:17)
    at require (internal/module.js:11:18)
    at Object.<anonymous> (/home/build/jenkins/workspace/sdk_titanium_mobile_PR-9773-5JN66CB7GZHYQRQBPRYCUFOT7LIU6U2ZROVNLTIB32CQWV2PUFKA/node_modules/babel-core/lib/util.js:39:18)

```

<p align="right">
  Generated by :no_entry_sign: <a href="http://github.com/danger/danger-js/">dangerJS</a>
</p>

sgtcoolguy commented on iphone/Classes/UtilsModule.m:
```diff
@@ -58,7 +58,17 @@ - (TiBlob *)base64decode:(id)args
+    [str appendString:@"="];
+  }
+
+  if ([str rangeOfString:@"="].length > 2) {
```
Wouldn't you want to check the string before adding padding?

sgtcoolguy commented on iphone/Classes/UtilsModule.m:
```diff
@@ -58,7 +58,17 @@ - (TiBlob *)base64decode:(id)args
+  NSMutableString *str = [[NSMutableString alloc] initWithString:[[self convertToString:args]
+                                                                     stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceAndNewlineCharacterSet]]];
+  // Fill padding for backwards compatibility
+  while ([str length] % 4 != 0) {
```
I don't think this is right. Base64 encoded strings will only have one or two '=' characters as padding, which suggest modulo 3, not 4.

https://stackoverflow.com/questions/4080988/why-does-base64-encoding-require-padding-if-the-input-length-is-not-divisible-by

hansemannn replied:
The RFC sounds helpful:

```
(1) the final quantum of encoding input is an integral multiple of 24
   bits; here, the final unit of encoded output will be an integral
   multiple of 4 characters with no "=" padding,

   (2) the final quantum of encoding input is exactly 8 bits; here, the
   final unit of encoded output will be two characters followed by two
   "=" padding characters, or

   (3) the final quantum of encoding input is exactly 16 bits; here, the
   final unit of encoded output will be three characters followed by one
   "=" padding character.
```

sgtcoolguy reviewed (changes requested):
You already have a manual example in the JIRA ticket, please add it to our Base64 unit tests.

Should be pretty easy to add a few tests for improperly padded encoded strings and verify we handle them and decode them as expected.

sgtcoolguy commented:
Some possible unit test examples:

Decoded | Properly padded | Improperly Padded
------------ | ------------- | ------------
f | Zg== | Zg
fo | Zm8= | Zm8
foo | Zm9v| *N/A*
foob | Zm9vYg== | Zm9vYg
fooba | Zm9vYmE= | Zm9vYmE
foobar | Zm9vYmFy | *N/A*

hansemannn commented:
So it's working..but one of them ('foo' <==> 'Zmb9v') fails for me.

```js
var win = Ti.UI.createWindow({
    backgroundColor: '#fff'
});

var btn = Ti.UI.createButton({
    title: 'Trigger'
});

btn.addEventListener('click', function() {
  // Should be a JSON-payload
  if (!Ti.Utils.base64decode('eyJzdWIiOiJTc2VsdmFyYWoiLCJlbWFpbCI6IlNzZWx2YXJhaiIsImF1ZCI6ImxhYmFsZXJ0IiwianRpIjoibG9Mczh3ajlpbFNTS1ZyRXFvOFowUyIsImlzcyI6Imh0dHBzOlwvXC9zc28uY2VsZ2VuZS5jb20iLCJpYXQiOjE1MTYxNzY0OTcsImV4cCI6MTUxNjE3Njc5NywicGkuc3JpIjoiTEk0ZjFvNUNqam81ejhQdVJOcWNreU95WVE4In0')) {
    alert('ERROR (1)!');
  }

  // More tests!
  if (Ti.Utils.base64decode('Zg').text !== 'f') {
    alert('ERROR (2)!');
  }
  if (Ti.Utils.base64decode('Zm8').text !== 'fo') {
    alert('ERROR (3)!');
  }
  if (Ti.Utils.base64decode('Zm9v').text != 'foo') {
    alert('ERROR (4)');
  }
  if (Ti.Utils.base64decode('Zm9vYg').text !== 'foob') {
    alert('ERROR (5)!');
  }
  if (Ti.Utils.base64decode('Zm9vYmE').text !== 'fooba') {
    alert('ERROR (6)!');
  }
  if (Ti.Utils.base64decode('Zm9vYmFy').text !== 'foobar') {
    alert('ERROR (7)!');
  }
});

win.add(btn);
win.open();
```

It obviously returns "foo", but debugging it returns a length of 4 with a "NULL" unicode char at the end. I have no idea where that comes from, but it blocks.

hansemannn commented:
@sgtcoolguy Fixed the iOS padding for all cases, verified the unit-tests and fixed a few more of them (all iOS and Android ones). iOS now accepts the "TiFile" parameter as well and Android does for the decoding as well (only worked for encoding on Android). Test-case is also listed on JIRA.

hansemannn committed:
[TIMOB-25727] Fix Base64 padding and unit-tests

hansemannn committed:
[TIMOB-25727] Handle Android IOException properly

build commented:
<!--
  1 failure:  Tests have failed...
  0 warning: 
  1 messages
  1 markdown notices
-->

<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Fails</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:no_entry_sign:</td>
      <td>

  Tests have failed, see below for more information.
  </td>
    </tr>
  </tbody>
</table>



<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Messages</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:book:</td>
      <td>

  :floppy_disk: <a href="https://jenkins.appcelerator.org/job/titanium-sdk/job/titanium_mobile/job/PR-9773/6/artifact/dist/mobilesdk-7.1.0.v20180129015820-osx.zip">Here's the generated SDK zipfile</a>.
  </td>
    </tr>
  </tbody>
</table>

### Tests: 

| Classname | Name | Time | Error |
| --- | --- | --- | --- |
| android.Titanium.Utils | #base64encode(Ti.Filesystem.File) | 0.001 | Error: the string "Not yet implemented. TYPE_STREAM_BASE64" was thrown, throw an Error :) |

<p align="right">
  Generated by :no_entry_sign: <a href="http://github.com/danger/danger-js/">dangerJS</a>
</p>

hansemannn committed:
[TIMOB-25727] Fix Android base64encode with Ti.File

# Conflicts:
#	android/modules/utils/src/java/ti/modules/titanium/utils/UtilsModule.java

sgtcoolguy commented on apidoc/Titanium/Utils/Utils.yml:
```diff
@@ -10,18 +10,19 @@ platforms: [android, iphone, ipad]
     description: |
-        Starting from Titanium 3.3.0, `obj` can be a [File](Titanium.Filesystem.File) object
-        on Android.
+        The parameter can be a String, Ti.Blob or Ti.File (iOS and Android only).
```
May still want to retain note that Ti.Filesystem/File is supported on Android 3.3+, iOS 7.0.3+ (or whatever version this PR and cherry-picks gets into).

sgtcoolguy commented on tests/Resources/ti.utils.test.js:
```diff
@@ -0,0 +1,135 @@
+	});
+
+	// FIXME Windows gives: 'base64decode: attempt to decode a value not in base64 char set'
+	it.windowsBroken('#base64decode(Ti.Blob)', function () {
```
Add tests for `#base64decode(Ti.Filesystem.File)`

sgtcoolguy commented on tests/Resources/ti.utils.test.js:
```diff
@@ -0,0 +1,135 @@
+		should(test.getText()).be.eql('dGVzdA==');
+	});
+
+	it('#base64encode(Ti.Blob)', function () {
```
Add test for `#base64encode(Ti.Filesystem.File)`

hansemannn replied:
Already in https://github.com/appcelerator/titanium_mobile/pull/9773/files#diff-6384f93cac05d6aea1061d799b0fe88aR67. Other test is added.

sgtcoolguy reviewed (approved):
Looks good. I'd add tests for base64encode/decode with a Ti.Filesystem.File argument as well for completeness (we already have Ti.Blob test variants, so just pass the file in rather than convert the file to a Ti.Blob first)

hansemannn committed:
[TIMOB-25727] Improve docs, add more unit-tests

ewieberappc commented:
FR Passed on 7_0_X branch

ewieberappc committed:
Merge branch 'master' into TIMOB-25727
